### PR TITLE
Spell out nested Self type in lint message

### LINKED ISSUE #78600: Improve error message for "return type cannot contain a projection or `Self`..."
Following the improvement for the diagnostic of the error reported in this issue https://github.com/rust-lang/rust/issues/69276, it would be nice (depending on how https://github.com/rust-lang/rust/issues/61949 is prioritized) to have the same error message when instead of returning `Self` the function returns `Result<Self, Error>`.

When `Self` is returned I get a nice error message (on rustc 1.47):

```rust
struct S<'a>(&'a i32);

impl<'a> S<'a> {
    async fn new(i: &'a i32) -> Self {
        S(&22)
    }
}
```

```
error[E0760]: `async fn` return type cannot contain a projection or `Self` that references lifetimes from a parent scope
 --> src/lib.rs:4:33
  |
4 |     async fn new(i: &'a i32) -> Self {
  |                                 ^^^^ help: consider spelling out the type instead: `S<'a>`
```

But if `Result<Self, Error>` is used I get:

```rust
struct S<'a>(&'a i32);

impl<'a> S<'a> {
    async fn new(i: &'a i32) -> Result<Self, ()> {
        Ok(S(&22))
    }
}
```
```
error[E0760]: `async fn` return type cannot contain a projection or `Self` that references lifetimes from a parent scope
 --> src/lib.rs:4:33
  |
4 |     async fn new(i: &'a i32) -> Result<Self, ()> {
  |                                 ^^^^^^^^^^^^^^^^
```

with no suggestion to spell the type.

[Link to Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9fe72633cfd564cea84b0e86523fb7d6)

### PR DESCRIPTION
**Pull request by csmoe**
Closes #78600
r? @tmandry

### CODE REVIEW

File: compiler/rustc_typeck/src/check/check.rs [hunk 1]
```diff
@@ -1590,3 +1591,31 @@ fn opaque_type_cycle_error(tcx: TyCtxt<'tcx>, def_id: LocalDefId, span: Span) {
...
+
+    fn nested_visit_map(&mut self) -> hir::intravisit::NestedVisitorMap<Self::Map> {
+        hir::intravisit::NestedVisitorMap::OnlyBodies(self.tcx.hir())
+    }
+
+    fn visit_ty(&mut self, arg: &'tcx hir::Ty<'tcx>) {
+        match arg.kind {
+            hir::TyKind::Path(hir::QPath::Resolved(None, path)) => match &path.segments {
+                [segment]
+                    if segment.res.map(|res| matches!(res, Res::SelfTy(_, _))).unwrap_or(false) =>
```
Comment (tmandry) [inline compiler/rustc_typeck/src/check/check.rs:1611]: nit: I think you could write this as

```suggestion
                [PathSegment { res: Res::SelfTy(_, _), .. }] => {
```

File: src/test/ui/async-await/issues/issue-78600.stderr [hunk 2]
```diff
@@ -0,0 +1,11 @@
+error[E0760]: `async fn` return type cannot contain a projection or `Self` that references lifetimes from a parent scope
+  --> $DIR/issue-78600.rs:6:33
+   |
+LL |     async fn new(i: &'a i32) -> Result<Self, ()> {
+   |                                 ^^^^^^^----^^^^^
+   |                                        |
+   |                                        help: consider spelling out the type instead: `std::result::Result<S<'a>, ()>`
```
Comment (tmandry) [inline src/test/ui/async-await/issues/issue-78600.stderr:7]: How hard would it be to only print the Self part of the type, i.e. `S<'a>`?

It seems like you might be able to save it in the `ProhibitOpaqueVisitor`. Might not work in every case.

**Comment by csmoe** [comment]
@tmandry thanks for your reviews, addressed :)

**Comment by tmandry** [comment]
Looks great, thanks @csmoe!

@bors r+

**Pull request merged**